pkg/adapter/awscloudwatchlogsource/adapter.go

```diff
@@ -135,38 +127,19 @@ func (a *adapter) Start(ctx context.Context) error {
 
 	// Setup polling to retrieve metrics
 	poll := time.NewTicker(a.pollingInterval)
-	metricsCh := make(chan bool)
 	defer poll.Stop()
-	defer close(metricsCh)
-
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	// Cleanup thread
-	go func() {
-		<-ctx.Done()
-		a.logger.Info("Shutdown signal received. Terminating")
-		metricsCh <- true
-		cancel()
-	}()
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-
-	// Primary execution thread. Wake up every pollintInterval, and retrieve the logs
-	go func() {
-		for {
-			select {
-			case <-metricsCh:
-				wg.Done()
-				return
-			case t := <-poll.C:
-				go a.CollectLogs(t)
-			}
+
+	// Wake up every pollingInterval, and retrieve the logs
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case t := <-poll.C:
+			a.CollectLogs(t)
```

Comment:
In this case, I opted to use the go-routine to ensure log and metric collection/event generation did not interfere with the polling interval.

Reply:
The ticker keeps running even if a function is blocking the select, unlike a timer. Is that what you're worried about?

Reply:
More like:
1. Ticker fires every 2 minutes
2. Processing takes longer than two minutes (suppose an extra 30 seconds)
3. On next loop, processing uses the current time, and the starting time interval from now - 2 minutes resulting in a loss of metric/log data of that delta.

Granted this could be worked around by passing the previous fire time with the current time into the collection functions, but this also raises a concern of a chain reaction delay where the processing of the first event delays the processing of the second event, and so on with the delays growing over time. The goroutine approach fans out the processing so the delays won't impact the primary server thread unless they grow in enough number to slow the overall system down and summon the thundering herd.

Reply:
The question here is: is it reasonable to run multiple `CollectLogs` in parallel? I'm reverting this change, but the design looks shaky to me, especially with this calculation based on the interval. If you don't base your startTime on the timestamp of the last collected log, a delayed tick will potentially cause missed entries. But that's off topic here.